File: frontend/src/features/communications/utils/segmentCounter.ts
/**
 * SMS segment counter — GSM-7 vs UCS-2 detection and segment calculation.
 *
 * Mirrors backend `services/sms/segment_counter.py` exactly.
 * Validates: Requirements 15.9, 34, 43
 */

// --- Constants ---

export const SENDER_PREFIX = "Grin's Irrigation: ";
export const STOP_FOOTER = ' Reply STOP to opt out.';
export const ALLOWED_MERGE_FIELDS = ['first_name', 'last_name', 'next_appointment_date'] as const;

// --- GSM-7 character sets ---

const GSM7_BASIC = new Set(
  '@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ ÆæßÉ' +
  ' !"#¤%&\'()*+,-./0123456789:;<=>?' +
  '¡ABCDEFGHIJKLMNOPQRSTUVWXYZ' +
  'ÄÖÑܧ¿abcdefghijklmnopqrstuvwxyz' +
  'äöñüà',
);

const GSM7_EXTENSION = new Set('^{}\\[~]|€');

export type Encoding = 'GSM-7' | 'UCS-2';

function detectEncoding(text: string): Encoding {
  for (const ch of text) {
    if (!GSM7_BASIC.has(ch) && !GSM7_EXTENSION.has(ch)) return 'UCS-2';
  }
  return 'GSM-7';
}

function gsm7CharCount(text: string): number {
  let count = 0;
  for (const ch of text) {
    count += GSM7_EXTENSION.has(ch) ? 2 : 1;
  }
  return count;
}

/** Count SMS segments — matches backend `segment_counter.count_segments`. */
export function countSegments(
  text: string,
  options?: { includePrefix?: boolean; includeFooter?: boolean },
): { encoding: Encoding; segments: number; chars: number } {
  const prefix = (options?.includePrefix ?? true) ? SENDER_PREFIX : '';
  const footer = (options?.includeFooter ?? true) ? STOP_FOOTER : '';
  const full = `${prefix}${text}${footer}`;

  const encoding = detectEncoding(full);

  let chars: number;
  let segments: number;
  if (encoding === 'GSM-7') {
    chars = gsm7CharCount(full);
    segments = chars <= 160 ? 1 : Math.ceil(chars / 153);
  } else {
    chars = full.length;
    segments = chars <= 70 ? 1 : Math.ceil(chars / 67);
  }

  return { encoding, segments, chars };
}

// --- Merge-field helpers ---

const MERGE_FIELD_RE = /\{(\w+)\}/g;

export function findInvalidMergeFields(text: string): string[] {
  const invalid: string[] = [];
  let match: RegExpExecArray | null;
  const re = new RegExp(MERGE_FIELD_RE.source, MERGE_FIELD_RE.flags);
  while ((match = re.exec(text)) !== null) {
    if (!(ALLOWED_MERGE_FIELDS as readonly string[]).includes(match[1])) {
      invalid.push(match[1]);
    }
  }
  return [...new Set(invalid)];
}

/** Render merge fields in text using recipient data. Missing keys → empty string. */
export function renderTemplate(text: string, context: Record<string, string>): string {
  return text.replace(/\{(\w+)\}/g, (_, key: string) => context[key] ?? '');
}

File: src/grins_platform/services/sms/segment_counter.py
"""SMS segment counter — GSM-7 vs UCS-2 detection and segment calculation.

Includes the auto-appended sender prefix and STOP footer in the count.
"""

from __future__ import annotations

import math
import os
from typing import Literal

# GSM-7 basic character set (standard 7-bit encoding)
_GSM7_BASIC: frozenset[str] = frozenset(
    "@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ ÆæßÉ"
    " !\"#¤%&'()*+,-./0123456789:;<=>?"
    "¡ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    "ÄÖÑܧ¿abcdefghijklmnopqrstuvwxyz"
    "äöñüà",
)

# GSM-7 extension characters (each costs 2 chars: escape + char)
_GSM7_EXTENSION: frozenset[str] = frozenset("^{}\\[~]|€")

Encoding = Literal["GSM-7", "UCS-2"]

# Defaults
_DEFAULT_PREFIX = "Grin's Irrigation: "
_DEFAULT_FOOTER = " Reply STOP to opt out."


def _detect_encoding(text: str) -> Encoding:
    """Return 'GSM-7' if all chars are in GSM-7 alphabet, else 'UCS-2'."""
    for ch in text:
        if ch not in _GSM7_BASIC and ch not in _GSM7_EXTENSION:
            return "UCS-2"
    return "GSM-7"


def _gsm7_char_count(text: str) -> int:
    """Count GSM-7 character units (extension chars cost 2)."""
    return sum(2 if ch in _GSM7_EXTENSION else 1 for ch in text)


def count_segments(
    text: str,
    *,
    include_prefix: bool = True,
    include_footer: bool = True,
) -> tuple[Encoding, int, int]:
    """Count SMS segments for *text* including prefix and STOP footer.

    Returns ``(encoding, segments, char_count)`` where *char_count* is
    the total character units (GSM-7 extension chars count as 2).
    """
    prefix = (
        os.environ.get("SMS_SENDER_PREFIX", _DEFAULT_PREFIX) if include_prefix else ""
    )
    footer = _DEFAULT_FOOTER if include_footer else ""
    full = f"{prefix}{text}{footer}"

    encoding = _detect_encoding(full)

    if encoding == "GSM-7":
        chars = _gsm7_char_count(full)
        segments = 1 if chars <= 160 else math.ceil(chars / 153)
    else:
        chars = len(full)
        segments = 1 if chars <= 70 else math.ceil(chars / 67)

    return encoding, segments, chars
